kasa/aestransport.py

```diff
@@ -243,96 +241,23 @@ def handshake_session_expired(self):
-                    raise SmartDeviceException(
-                        f"Unable to connect to the device: {self.host}: {sdex}"
-                    ) from sdex
-                continue
-            except httpx.ConnectError as cex:
-                await self.close()
-                raise SmartDeviceException(
-                    f"Unable to connect to the device: {self.host}: {cex}"
-                ) from cex
-            except TimeoutError as tex:
-                await self.close()
-                raise SmartDeviceException(
-                    f"Unable to connect to the device, timed out: {self.host}: {tex}"
-                ) from tex
-            except AuthenticationException as auex:
-                _LOGGER.debug("Unable to authenticate with %s, not retrying", self.host)
-                raise auex
-            except Exception as ex:
-                await self.close()
-                if retry >= retry_count:
-                    _LOGGER.debug("Giving up on %s after %s retries", self.host, retry)
-                    raise SmartDeviceException(
-                        f"Unable to connect to the device: {self.host}: {ex}"
-                    ) from ex
-                continue
-
-        # make mypy happy, this should never be reached..
-        raise SmartDeviceException("Query reached somehow to unreachable")
-
-    async def _execute_query(self, request: Union[str, Dict], retry_count: int) -> Dict:
-        _LOGGER.debug(
-            "%s >> %s",
-            self.host,
-            _LOGGER.isEnabledFor(logging.DEBUG) and pf(request),
-        )
-
-        if not self.http_client:
-            self.http_client = httpx.AsyncClient()
-
-        if not self.handshake_done or self.handshake_session_expired():
-            try:
-                await self.perform_handshake()
-                await self.perform_login(False)
-            except AuthenticationException:
-                await self.perform_handshake()
-                await self.perform_login(True)
-
-        if isinstance(request, dict):
-            aes_method = next(iter(request))
-            aes_params = request[aes_method]
-        else:
-            aes_method = request
-            aes_params = None
-
-        aes_request = self.get_aes_request(aes_method, aes_params)
-        response_data = await self.send_secure_passthrough(aes_request)
-
-        _LOGGER.debug(
-            "%s << %s",
-            self.host,
-            _LOGGER.isEnabledFor(logging.DEBUG) and pf(response_data),
-        )
-
-        return response_data
+    async def send(self, request: str):
+        """Send the request."""
+        if self.needs_handshake():
+            raise SmartDeviceException(
+                "Handshake must be complete before trying to send"
+            )
+        if self.needs_login():
+            raise SmartDeviceException("Login must be complete before trying to send")
```

Comment:
Just a question, would it make sense to perform handshaking (and login, too?) transparently whenever needed?